docs/src/prototypes/MenuButton/MenuButton.tsx

```diff
@@ -0,0 +1,198 @@
+
+  componentDidUpdate(_, prevState: MenuButtonState) {
+    if (!prevState.menuOpen && this.state.menuOpen) {
+      document.addEventListener('click', this.handleDocumentClick)
+
+      focusMenuItem(this.menuRef, this.state.lastKeyCode === keyboardKey.ArrowUp ? 'last' : 'first')
+    }
+
+    if (prevState.menuOpen && !this.state.menuOpen) {
+      document.removeEventListener('click', this.handleDocumentClick)
+
+      switch (this.state.lastKeyCode) {
+        case keyboardKey.Enter:
+        case keyboardKey.Escape:
+          focusButton(this.buttonRef)
+          break
+
+        case keyboardKey.Tab:
+          focusNearest(this.buttonRef, this.state.lastShiftKey ? 'previous' : 'next')
+          break
+      }
+    }
+  }
+
+  handleButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.setState(prevState => ({
+        lastKeyCode: null,
+        menuOpen: !prevState.menuOpen,
+      }))
+    },
+  })
+
+  handleDocumentClick = (e: MouseEvent) => {
+    const { menuOpen } = this.state
+    const target = e.target as HTMLElement
+    const isInside =
+      _.invoke(this.buttonRef.current, 'contains', target) &&
+      _.invoke(this.menuRef.current, 'contains', target)
+
+    if (menuOpen && !isInside) {
+      this.setState({ lastKeyCode: null, menuOpen: false })
+    }
+  }
+
+  handleKeyDown = (e: React.KeyboardEvent) => {
+    const { menuOpen } = this.state
+
+    const keyCode = keyboardKey.getCode(e)
+    const shouldClose =
+      menuOpen && _.includes([keyboardKey.Enter, keyboardKey.Escape, keyboardKey.Tab], keyCode)
+    const shouldOpen =
+      !menuOpen && _.includes([keyboardKey.ArrowDown, keyboardKey.ArrowUp], keyCode)
+
+    if (shouldClose) {
+      // We should prevent default there otherwise event will be bubbled to a button and will cause menu reopen
+      if (keyCode === keyboardKey.Enter) e.preventDefault()
+
+      this.setState({
+        lastKeyCode: keyCode,
+        lastShiftKey: e.shiftKey,
+        menuOpen: false,
+      })
+    }
+
+    if (shouldOpen) {
+      this.setState({ lastKeyCode: keyCode, menuOpen: true })
+    }
+  }
+
+  handleMenuItemOverrides = accessibilityBehavior =>
+    _.map(_.get(this.props.menu, 'items'), (item: ShorthandValue) =>
+      typeof item === 'object'
+        ? {
+            ...item,
+            ...accessibilityBehavior.menuItem,
+          }
+        : {
+            content: item,
+            key: item,
+            ...accessibilityBehavior.menuItem,
+          },
+    )
+
+  render() {
+    const { accessibility, button, disabled, menu, placement } = this.props
+    const { menuOpen } = this.state
+    const accessibilityBehavior: AccessibilityBehavior = accessibility({
+      ...this.props,
+      ...this.state,
+    })
+
+    return (
+      <MenuButtonWrapper onKeyDown={this.handleKeyDown}>
```

Comment:
just as a matter of thought - should we, instead, remove this `MenuButtonWrapper` component and make `MenuButton` to be the one that is created as Stardust component? I mean, I do understand that the fact `state` is used in `MenuButton`, potentially, had forced towards using wrapper as Stardust component, but now 
- we are defining accessibility as a prop of `MenuButton`  component
- and instead of letting Stardust to process the accessibility, we are using this internal logic here

I would expect things top remain encapsulated - so that, in case if there will be any update to the way how `accessibility` is processed, this update will reach `MenuButton` component as well.

What do you think?

Reply:
Yep, I removed `MenuButtonWrapper` as it doesn't make any reasonable value, but I don't see any way to make `MenuButton` as Stardust component except the HOC.